backend/quatrix/quatrix.go

```diff
@@ -0,0 +1,1255 @@
+func (o *Object) Open(ctx context.Context, options ...fs.OpenOption) (in io.ReadCloser, err error) {
+	if o.id == "" {
+		return nil, errors.New("can't download - no id")
+	}
+
+	linkID, err := o.fs.downloadLink(ctx, o.id)
+	if err != nil {
+		return nil, err
+	}
+
+	fs.FixRangeOption(options, o.size)
+
+	opts := rest.Opts{
+		Method:  "GET",
+		Path:    "/file/download/" + linkID,
+		Options: options,
+	}
+
+	var resp *http.Response
+
+	err = o.fs.pacer.Call(func() (bool, error) {
+		resp, err = o.fs.srv.Call(ctx, &opts)
+		return shouldRetry(ctx, resp, err)
+	})
+	if err != nil {
+		return nil, err
+	}
+	return resp.Body, err
+}
+
+func (f *Fs) downloadLink(ctx context.Context, id string) (linkID string, err error) {
+	linkParams := &api.IDList{
+		IDs: []string{id},
+	}
+	opts := rest.Opts{
+		Method: "POST",
+		Path:   "file/download-link",
+	}
+
+	var resp *http.Response
+	link := &api.DownloadLinkResponse{}
+
+	err = f.pacer.Call(func() (bool, error) {
+		resp, err = f.srv.CallJSON(ctx, &opts, linkParams, &link)
+		return shouldRetry(ctx, resp, err)
+	})
+	if err != nil {
+		return "", err
+	}
+	return link.ID, nil
+}
+
+// SetModTime sets the modification time of the local fs object
+func (o *Object) SetModTime(ctx context.Context, t time.Time) error {
+	file, err := o.fs.setMTime(ctx, o.id, t)
+	if err != nil {
+		return fmt.Errorf("set mtime: %w", err)
+	}
+
+	return o.setMetaData(file)
+}
+
+// Update the object with the contents of the io.Reader, modTime and size
+//
+// The new object may have been created if an error is returned
+func (o *Object) Update(ctx context.Context, in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) (err error) {
+	size := src.Size()
+	modTime := src.ModTime(ctx)
+	remote := o.Remote()
+
+	// Create the directory for the object if it doesn't exist
+	leaf, directoryID, err := o.fs.dirCache.FindPath(ctx, remote, true)
+	if err != nil {
+		return err
+	}
+
+	uploadSession, err := o.uploadSession(ctx, directoryID, leaf)
+	if err != nil {
+		return fmt.Errorf("object update: %w", err)
+	}
+
+	o.id = uploadSession.FileID
+
+	defer func() {
+		if err == nil {
+			return
+		}
+
+		deleteErr := o.fs.deleteObject(ctx, o.id)
+		if deleteErr != nil {
+			fs.Logf(o.remote, "remove: %s", deleteErr)
+		}
+	}()
+
+	return o.dynamicUpload(ctx, size, modTime, in, uploadSession, options...)
+}
+
+// dynamicUpload uploads object in chunks, which are being dynamically recalculated on each iteration
+// depending on upload speed in order to make upload faster
+func (o *Object) dynamicUpload(ctx context.Context, size int64, modTime time.Time, in io.Reader,
+	uploadSession *api.UploadLinkResponse, options ...fs.OpenOption) error {
+	var (
+		speed      float64
+		localChunk int64
+	)
+
+	defer o.fs.uploadMemoryManager.Return(o.id)
+
+	for offset := int64(0); offset < size; offset += localChunk {
+		localChunk = o.fs.uploadMemoryManager.Consume(o.id, size-offset, speed)
+
+		buf := make([]byte, localChunk)
```

Comment:
Rather than making `buf` you'd be better off using `lib/multipart.NewRW()` which you can use as a memory buffer but rclone will recycle the memory and/or use mmap to allocate it.